cpp/tests/utilities/column_utilities.cu

```diff
@@ -309,47 +549,100 @@ struct column_comparator_impl<list_view, check_exact_equality> {
         return mask == nullptr ? true : cudf::bit_is_set(mask, index + offset);
       });
 
-    auto diff_iter = thrust::copy_if(
+    // when checking for equivalency, we can't compare offset values directly, we can only
+    // compare lengths of the rows, and only if valid.  as a concrete example, you could have two
+    // equivalent columns with the following data:
+    //
+    // column A
+    //    offsets =  [0, 3, 5, 7]
+    //    validity = [0, 1, 1, 1]
+    //
+    // column B
+    //   offsets =   [0, 0, 2, 4]
+    //   validity =  [0, 1, 1, 1]
+    //
+    // Row 0 in column A happens to have a positive length, even though the row is null, but column
+    // B does not.  So the offsets for the remaining valid rows are fundamentally different even
+    // though the row lengths are the same.
+    //
+    auto input_iter = thrust::make_counting_iterator(0);
+    auto diff_iter  = thrust::copy_if(
       rmm::exec_policy(),
-      thrust::make_counting_iterator(0),
-      thrust::make_counting_iterator(lhs_l.size() + 1),
+      input_iter,
+      input_iter + lhs_row_indices.size(),
       differences.begin(),
-      [lhs_offsets, rhs_offsets, lhs_valids, rhs_valids, num_rows = lhs_l.size()] __device__(
-        size_type index) {
-        // last offset has no validity associated with it
-        if (index < num_rows - 1) {
-          if (lhs_valids[index] != rhs_valids[index]) { return true; }
-          // if validity matches -and- is false, we can ignore the actual values. this
-          // is technically not checking "equal()", but it's how the non-list code path handles it
-          if (!lhs_valids[index]) { return false; }
+      [lhs_offsets,
+       rhs_offsets,
+       lhs_valids,
+       rhs_valids,
+       lhs_indices = lhs_row_indices.begin<size_type>(),
+       rhs_indices = rhs_row_indices.begin<size_type>()] __device__(size_type index) {
+        auto const lhs_index = lhs_indices[index];
+        auto const rhs_index = rhs_indices[index];
+
+        // check for validity match
+        if (lhs_valids[lhs_index] != rhs_valids[rhs_index]) { return true; }
+
+        // if the row is valid, check that the length of the list is the same. do this
+        // for both the equivalency and exact equality checks.
+        if (lhs_valids[lhs_index] && ((lhs_offsets[lhs_index + 1] - lhs_offsets[lhs_index]) !=
```

Comment:
I wonder if we can do something simpler, like pulling all contents of the original columns (using gather with gather map is the simple sequence `[0, size)`). Then, we just use the existing code to compare the resulted columns (which have zero size for the nulls).

Reply:
This is a bit of a large change, and one think I don't particularly like about it is that it uses a function (`gather()`) which will in turn be using _this_ function to verify itself.